cupyx/scipy/ndimage/_filters_core.py

```diff
@@ -0,0 +1,368 @@
+    return origin
+
+
+def _origins_to_offsets(origins, w_shape):
+    return tuple(x//2+o for x, o in zip(w_shape, origins))
+
+
+def _check_mode(mode):
+    if mode not in {'reflect', 'constant', 'nearest', 'mirror', 'wrap'}:
+        msg = 'boundary mode not supported (actual: {})'.format(mode)
+        raise RuntimeError(msg)
+    return mode
+
+
+def _get_inttype(input):
+    # The integer type to use for indices in the input array
+    # The indices actually use byte positions and we can't just use
+    # input.nbytes since that won't tell us the number of bytes between the
+    # first and last elements when the array is non-contiguous
+    nbytes = sum((x-1)*abs(stride) for x, stride in
+                 zip(input.shape, input.strides)) + input.dtype.itemsize
+    return 'int' if nbytes < (1 << 31) else 'ptrdiff_t'
+
+
+def _check_size_footprint_structure(ndim, size, footprint, structure,
+                                    stacklevel=3, force_footprint=False):
+    if structure is None and footprint is None:
+        if size is None:
+            raise RuntimeError("no footprint or filter size provided")
+        sizes = _fix_sequence_arg(size, ndim, 'size', int)
+        if force_footprint:
+            return None, cupy.ones(sizes, bool), None
+        return sizes, None, None
+    if size is not None:
+        import warnings
+        warnings.warn("ignoring size because {} is set".format(
+            'structure' if footprint is None else 'footprint'),
+            UserWarning, stacklevel=stacklevel+1)
+
+    if footprint is not None:
+        footprint = cupy.array(footprint, bool, True, 'C')
+        if not footprint.any():
+            raise ValueError("all-zero footprint is not supported")
+
+    if structure is None:
+        if not force_footprint and footprint.all():
+            if footprint.ndim != ndim:
+                raise RuntimeError("size must have length equal to input rank")
+            return footprint.shape, None, None
+        return None, footprint, None
+
+    structure = cupy.ascontiguousarray(structure)
+    if footprint is None:
+        footprint = cupy.ones(structure.shape, bool)
+    return None, footprint, structure
+
+
+def _convert_1d_args(ndim, weights, origin, axis):
+    if weights.ndim != 1 or weights.size < 1:
+        raise RuntimeError('incorrect filter size')
+    axis = cupy.util._normalize_axis_index(axis, ndim)
+    w_shape = [1]*ndim
+    w_shape[axis] = weights.size
+    weights = weights.reshape(w_shape)
+    origins = [0]*ndim
+    origins[axis] = _check_origin(origin, weights.size)
+    return weights, tuple(origins)
+
+
+def _check_nd_args(input, weights, mode, origin, wghts_name='filter weights'):
+    if input.dtype.kind == 'c':
+        raise TypeError('Complex type not supported')
+    _check_mode(mode)
+    # Weights must always be less than 2 GiB
+    if weights.nbytes >= (1 << 31):
+        raise RuntimeError('weights must be 2 GiB or less, use FFTs instead')
```

Comment:
Just curious, is it a simple rule of thumb, or is it something well-known to the image processing community?

Reply:
I think this was just a hard requirement because the kernels currently use 32-bit ints for indexing the weights. The cases where the `ndimage` filters perform well are when the `weights` kernel is much smaller than the image itself. That is a common usage scenario, so this is a reasonable assumption. For cases where the assumption does not hold, one should probably be using FFT-based filtering instead.

Reply:
grlee77 is correct, this was an optimization made for the by far most common (possibly always) situation that grlee77 points out. This does only apply to weights (the smaller of the two inputs), the other input is only restricted by available memory. The optimization probably doesn't have a huge impact on speed, so I could lift it if desired.

Reply:
Ah, I see. I missed the 32-bit indexing thing. Thanks for jumping in @grlee77 @coderforlife 🙂